src/components/DayPicker/DayPicker.tsx

```diff
@@ -270,22 +257,32 @@ export const DayPicker: React.FC<DayPickerProps> = ({
           [classes.textFieldNoManualInput]: !parseDate,
         })}
         onChange={_onTextChange}
-        onFocus={(e) => {
-          if (parseDate) {
-            setIsCalendarOpen(true);
+        onMouseDown={(e) => {
+          /**
+           * This handler allows for opening _and_ closing the popover via click in
+           * readonly mode.
+           *
+           * We need to use onMouseDown instead of onClick so we can run this _before_
+           * onFocus runs and avoid an immediate open/close.
+           */
+          if (!parseDate) {
+            setIsCalendarOpen((current) => !current);
           }
+
+          textFieldProps.onMouseDown?.(e);
+        }}
+        onFocus={(e) => {
+          setIsCalendarOpen(true);
           textFieldProps.onFocus?.(e);
         }}
-        /**
-         * Using onClick for readonly mode makes for a better experience!
-         *
-         * Primarily, it allows for closing the "popover" by clicking on the field again.
-         */
-        onClick={(e) => {
-          if (!parseDate) {
-            setIsCalendarOpen((current) => !current);
+        onBlur={(e) => {
+          /**
+           * This logic primarily serves to provide good keyboard navigation.
+           */
+          if (!isInternalNode(e.relatedTarget)) {
+            setIsCalendarOpen(false);
           }
-          textFieldProps.onClick?.(e);
+          textFieldProps.onBlur?.(e);
```

Comment:
This "pass-through" behavior is also used for a few props in `Modal`. Planning to PR a consolidation in the form of a `composeEventHandlers` helper in a subsequent PR.